File: 01.SERVIDOR/ws_comercializadora_gr08/src/main/java/ec/edu/monster/service/VentaService.java
package ec.edu.monster.service;

import ec.edu.monster.db.DatabaseConnection;
import ec.edu.monster.dto.*;
import ec.edu.monster.model.DetalleFactura;
import ec.edu.monster.model.Electrodomestico;
import ec.edu.monster.model.Factura;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

public class VentaService {
    
    private static final Logger LOGGER = Logger.getLogger(VentaService.class.getName());
    private static final String BANQUITO_API_URL = "http://localhost:8080/ws_banquito_gr08/api";
    private static final BigDecimal DESCUENTO_EFECTIVO = new BigDecimal("0.33");
    
    public VentaResponse procesarVenta(VentaRequest request) {
        VentaResponse response = new VentaResponse();
        
        try (Connection connection = DatabaseConnection.getConnection()) {
            connection.setAutoCommit(false);
            
            // Validar forma de pago
            if (!"EFECTIVO".equalsIgnoreCase(request.getFormaPago()) && !"CREDITO_DIRECTO".equalsIgnoreCase(request.getFormaPago())) {
                response.setVentaExitosa(false);
                response.setMensaje("Forma de pago no válida. Use EFECTIVO o CREDITO_DIRECTO");
                return response;
            }
            
            // Obtener información de electrodomésticos y calcular totales
            List<Electrodomestico> electrodomesticos = new ArrayList<>();
            BigDecimal subtotal = BigDecimal.ZERO;
            
            for (DetalleVentaRequest detalle : request.getDetalles()) {
                Electrodomestico electrodomestico = buscarElectrodomestico(connection, Integer.parseInt(detalle.getIdElectrodomestico()));
                if (electrodomestico == null) {
                    connection.rollback();
                    response.setVentaExitosa(false);
                    response.setMensaje("Electrodoméstico con ID " + detalle.getIdElectrodomestico() + " no encontrado");
                    return response;
                }
                
                electrodomesticos.add(electrodomestico);
                BigDecimal subtotalLinea = electrodomestico.getPrecioVenta().multiply(new BigDecimal(detalle.getCantidad()));
                subtotal = subtotal.add(subtotalLinea);
            }
            
            // Calcular descuento y total
            BigDecimal descuento = BigDecimal.ZERO;
            BigDecimal total = subtotal;
            
            if ("EFECTIVO".equalsIgnoreCase(request.getFormaPago())) {
                descuento = subtotal.multiply(DESCUENTO_EFECTIVO).setScale(2, RoundingMode.HALF_UP);
                total = subtotal.subtract(descuento);
            }
            
            // Verificar crédito directo si aplica
            if ("CREDITO_DIRECTO".equalsIgnoreCase(request.getFormaPago())) {
                boolean creditoAprobado = verificarCreditoBanquito(request.getCedulaCliente(), total, request.getPlazoMeses());
                if (!creditoAprobado) {
                    connection.rollback();
                    response.setVentaExitosa(false);
                    response.setMensaje("Crédito no aprobado por el banco Banquito");
                    return response;
                }
            }
            
            // Crear o actualizar cliente
            crearOActualizarCliente(connection, request.getCedulaCliente(), request.getNombreCliente());
            
            // Crear factura
            Integer idFactura = crearFactura(connection, request, subtotal, descuento, total);
            
            // Crear detalles de factura
            List<DetalleVentaResponse> detallesResponse = new ArrayList<>();
            for (int i = 0; i < request.getDetalles().size(); i++) {
                DetalleVentaRequest detalleRequest = request.getDetalles().get(i);
                Electrodomestico electrodomestico = electrodomesticos.get(i);
                
                BigDecimal subtotalLinea = electrodomestico.getPrecioVenta().multiply(new BigDecimal(detalleRequest.getCantidad()));
                
                crearDetalleFactura(connection, idFactura, detalleRequest.getIdElectrodomestico(), 
                                 detalleRequest.getCantidad(), electrodomestico.getPrecioVenta(), subtotalLinea);
                
                detallesResponse.add(new DetalleVentaResponse(
                    detalleRequest.getIdElectrodomestico(),
                    electrodomestico.getNombre(),
                    detalleRequest.getCantidad(),
                    electrodomestico.getPrecioVenta(),
                    subtotalLinea
                ));
            }
            
            connection.commit();
            
            // Configurar respuesta
            response.setVentaExitosa(true);
            response.setMensaje("Venta procesada exitosamente");
            response.setIdFactura(idFactura);
            response.setSubtotal(subtotal);
            response.setDescuento(descuento);
            response.setTotal(total);
            response.setFormaPago(request.getFormaPago());
            
            // Establecer estado según forma de pago
            if ("EFECTIVO".equalsIgnoreCase(request.getFormaPago())) {
                response.setEstadoFactura("PAGADA");
            } else if ("CREDITO_DIRECTO".equalsIgnoreCase(request.getFormaPago())) {
                response.setEstadoFactura("APROBADA");
            }
            
            response.setDetalles(detallesResponse);
            
        } catch (SQLException e) {
            LOGGER.log(Level.SEVERE, "Error procesando venta", e);
            response.setVentaExitosa(false);
            response.setMensaje("Error en el proceso de venta");
        }
        
        return response;
    }
    
    private Electrodomestico buscarElectrodomestico(Connection connection, Integer idElectrodomestico) throws SQLException {
        String sql = "SELECT * FROM ELECTRODOMESTICO WHERE idElectrodomestico = ?";
        try (PreparedStatement stmt = connection.prepareStatement(sql)) {
            stmt.setInt(1, idElectrodomestico);
            ResultSet rs = stmt.executeQuery();
            if (rs.next()) {
                Electrodomestico electrodomestico = new Electrodomestico();
                electrodomestico.setIdElectrodomestico(rs.getInt("idElectrodomestico"));
                electrodomestico.setNombre(rs.getString("nombre"));
                electrodomestico.setPrecioVenta(rs.getBigDecimal("precioVenta"));
                return electrodomestico;
            }
        }
        return null;
    }
    
    private boolean verificarCreditoBanquito(String cedula, BigDecimal monto, Integer plazoMeses) {
        try {
            // Llamar al servicio del banco Banquito
            java.net.URL url = new java.net.URL(BANQUITO_API_URL + "/creditos/evaluar");
            java.net.HttpURLConnection conn = (java.net.HttpURLConnection) url.openConnection();
            conn.setRequestMethod("POST");
            conn.setRequestProperty("Content-Type", "application/json");
            conn.setDoOutput(true);
            
            // Construir JSON request
            String jsonInputString = String.format(
                "{\"cedula\":\"%s\",\"montoElectrodomestico\":%s,\"plazoMeses\":%d}",
                cedula, monto.toString().replace(',', '.'), plazoMeses != null ? plazoMeses : 12
            );
            
            LOGGER.log(Level.INFO, "Enviando solicitud a Banquito: {0}", jsonInputString);
            
            try (java.io.OutputStream os = conn.getOutputStream()) {
                byte[] input = jsonInputString.getBytes("utf-8");
                os.write(input, 0, input.length);
            }
            
            int responseCode = conn.getResponseCode();
            LOGGER.log(Level.INFO, "Response Code de Banquito: {0}", responseCode);
            
            if (responseCode == 200) {
                try (java.io.BufferedReader br = new java.io.BufferedReader(
                        new java.io.InputStreamReader(conn.getInputStream(), "utf-8"))) {
                    StringBuilder response = new StringBuilder();
                    String responseLine;
                    while ((responseLine = br.readLine()) != null) {
                        response.append(responseLine.trim());
                    }
                    
                    // Simple JSON parsing para verificar si el crédito fue aprobado
                    String jsonResponse = response.toString();
                    LOGGER.log(Level.INFO, "Respuesta de Banquito: {0}", jsonResponse);
                    
                    boolean aprobado = jsonResponse.contains("\"creditoAprobado\":true");
                    LOGGER.log(Level.INFO, "Crédito aprobado: {0}", aprobado);
                    
                    return aprobado;
                }
            }
            
        } catch (Exception e) {
            LOGGER.log(Level.WARNING, "Error verificando crédito con Banquito", e);
        }
        
        return false;
    }
    
    private void crearOActualizarCliente(Connection connection, String cedula, String nombre) throws SQLException {
        String checkSql = "SELECT COUNT(*) FROM CLIENTE WHERE cedula = ?";
        try (PreparedStatement checkStmt = connection.prepareStatement(checkSql)) {
            checkStmt.setString(1, cedula);
            ResultSet rs = checkStmt.executeQuery();
            if (rs.next() && rs.getInt(1) == 0) {
                // Insertar nuevo cliente
                String insertSql = "INSERT INTO CLIENTE (cedula, nombre) VALUES (?, ?)";
                try (PreparedStatement insertStmt = connection.prepareStatement(insertSql)) {
                    insertStmt.setString(1, cedula);
                    insertStmt.setString(2, nombre);
                    insertStmt.executeUpdate();
                }
            }
        }
    }
    
    private Integer crearFactura(Connection connection, VentaRequest request, BigDecimal subtotal, BigDecimal descuento, BigDecimal total) throws SQLException {
        // Determinar estado según forma de pago
        String estadoFactura = "EFECTIVO".equalsIgnoreCase(request.getFormaPago()) ? "PAGADA" : "APROBADA";
        
        String sql = "INSERT INTO FACTURA (cedula, fecha, formaPago, subtotal, descuento, total, estado) " +
                    "VALUES (?, CURDATE(), ?, ?, ?, ?, ?)";
        try (PreparedStatement stmt = connection.prepareStatement(sql, PreparedStatement.RETURN_GENERATED_KEYS)) {
            stmt.setString(1, request.getCedulaCliente());
            stmt.setString(2, request.getFormaPago());
            stmt.setBigDecimal(3, subtotal);
            stmt.setBigDecimal(4, descuento);
            stmt.setBigDecimal(5, total);
            stmt.setString(6, estadoFactura);
            stmt.executeUpdate();
            
            ResultSet rs = stmt.getGeneratedKeys();
            if (rs.next()) {
                return rs.getInt(1);
            }
        }
        return null;
    }
    
    private void crearDetalleFactura(Connection connection, Integer idFactura, String idElectrodomestico, 
                                   String cantidad, BigDecimal precioUnitario, BigDecimal subtotalLinea) throws SQLException {
        String sql = "INSERT INTO DETALLE_FACTURA (idFactura, idElectrodomestico, cantidad, precioUnitario, subtotalLinea) " +
                    "VALUES (?, ?, ?, ?, ?)";
        try (PreparedStatement stmt = connection.prepareStatement(sql)) {
            stmt.setInt(1, idFactura);
            stmt.setInt(2, Integer.parseInt(idElectrodomestico));
            stmt.setInt(3, Integer.parseInt(cantidad));
            stmt.setBigDecimal(4, precioUnitario);
            stmt.setBigDecimal(5, subtotalLinea);
            stmt.executeUpdate();
        }
    }
    
    public List<Electrodomestico> listarElectrodomesticos() {
        List<Electrodomestico> electrodomesticos = new ArrayList<>();
        
        try (Connection connection = DatabaseConnection.getConnection()) {
            String sql = "SELECT * FROM ELECTRODOMESTICO ORDER BY nombre";
            try (PreparedStatement stmt = connection.prepareStatement(sql)) {
                ResultSet rs = stmt.executeQuery();
                while (rs.next()) {
                    Electrodomestico electrodomestico = new Electrodomestico();
                    electrodomestico.setIdElectrodomestico(rs.getInt("idElectrodomestico"));
                    electrodomestico.setNombre(rs.getString("nombre"));
                    electrodomestico.setPrecioVenta(rs.getBigDecimal("precioVenta"));
                    electrodomesticos.add(electrodomestico);
                }
            }
        } catch (SQLException e) {
            LOGGER.log(Level.SEVERE, "Error listando electrodomésticos", e);
        }
        
        return electrodomesticos;
    }
}

File: SERVIDOR SOAP/ws_comer_dotnet_gr08/ec.edu.monster.service/VentaService.cs
using System;
using System.Collections.Generic;
using System.Data.SqlClient;
using System.ServiceModel;
using ec.edu.monster.db;
using ec.edu.monster.model;
using ws_banquito_dotnet_gr08;
using ws_comer_dotnet_gr08;
using BanquitoService = ws_banquito_dotnet_gr08.IService1;

namespace ec.edu.monster.service
{
    public class VentaService
    {
        private static readonly decimal DESCUENTO_EFECTIVO = 0.33m;
        private const string BANQUITO_SOAP_URL = "http://localhost:8733/Design_Time_Addresses/ws_banquito_dotnet_gr08/Service1/";

        public VentaResponse ProcesarVenta(VentaRequest request)
        {
            var response = new VentaResponse();

            using (var connection = DatabaseConnection.GetConnection())
            using (var transaction = connection.BeginTransaction())
            {
                try
                {
                    if (!string.Equals(request.FormaPago, "EFECTIVO", StringComparison.OrdinalIgnoreCase) &&
                        !string.Equals(request.FormaPago, "CREDITO_DIRECTO", StringComparison.OrdinalIgnoreCase))
                    {
                        response.VentaExitosa = false;
                        response.Mensaje = "Forma de pago no válida. Use EFECTIVO o CREDITO_DIRECTO";
                        return response;
                    }

                    var electrodomesticos = new List<Electrodomestico>();
                    decimal subtotal = 0m;

                    foreach (var detalle in request.Detalles)
                    {
                        var electrodomestico = BuscarElectrodomestico(connection, transaction, int.Parse(detalle.IdElectrodomestico));
                        if (electrodomestico == null)
                        {
                            transaction.Rollback();
                            response.VentaExitosa = false;
                            response.Mensaje = "Electrodoméstico con ID " + detalle.IdElectrodomestico + " no encontrado";
                            return response;
                        }

                        electrodomesticos.Add(electrodomestico);
                        var subtotalLinea = electrodomestico.PrecioVenta * decimal.Parse(detalle.Cantidad);
                        subtotal += subtotalLinea;
                    }

                    decimal descuento = 0m;
                    decimal total = subtotal;

                    if (string.Equals(request.FormaPago, "EFECTIVO", StringComparison.OrdinalIgnoreCase))
                    {
                        descuento = Math.Round(subtotal * DESCUENTO_EFECTIVO, 2, MidpointRounding.AwayFromZero);
                        total = subtotal - descuento;
                    }

                    if (string.Equals(request.FormaPago, "CREDITO_DIRECTO", StringComparison.OrdinalIgnoreCase))
                    {
                        var plazo = request.PlazoMeses ?? 12;
                        var creditoAprobado = VerificarCreditoBanquito(request.CedulaCliente, total, plazo);
                        if (!creditoAprobado)
                        {
                            transaction.Rollback();
                            response.VentaExitosa = false;
                            response.Mensaje = "Crédito no aprobado por el banco Banquito";
                            return response;
                        }
                    }

                    CrearOActualizarCliente(connection, transaction, request.CedulaCliente, request.NombreCliente);

                    var idFactura = CrearFactura(connection, transaction, request, subtotal, descuento, total);
                    if (!idFactura.HasValue)
                    {
                        throw new Exception("No se pudo obtener el identificador de la factura creada");
                    }

                    var detallesResponse = new List<DetalleVentaResponse>();
                    for (int i = 0; i < request.Detalles.Count; i++)
                    {
                        var detalleRequest = request.Detalles[i];
                        var electrodomestico = electrodomesticos[i];
                        var subtotalLinea = electrodomestico.PrecioVenta * decimal.Parse(detalleRequest.Cantidad);

                        CrearDetalleFactura(connection, transaction, idFactura.Value, detalleRequest.IdElectrodomestico, detalleRequest.Cantidad, electrodomestico.PrecioVenta, subtotalLinea);

                        detallesResponse.Add(new DetalleVentaResponse
                        {
                            IdElectrodomestico = detalleRequest.IdElectrodomestico,
                            NombreElectrodomestico = electrodomestico.Nombre,
                            Cantidad = detalleRequest.Cantidad,
                            PrecioUnitario = electrodomestico.PrecioVenta,
                            SubtotalLinea = subtotalLinea
                        });
                    }

                    transaction.Commit();

                    response.VentaExitosa = true;
                    response.Mensaje = "Venta procesada exitosamente";
                    response.IdFactura = idFactura;
                    response.Subtotal = subtotal;
                    response.Descuento = descuento;
                    response.Total = total;
                    response.FormaPago = request.FormaPago;
                    response.EstadoFactura = string.Equals(request.FormaPago, "EFECTIVO", StringComparison.OrdinalIgnoreCase)
                        ? "PAGADA"
                        : "APROBADA";
                    response.Detalles = detallesResponse;
                }
                catch (Exception ex)
                {
                    transaction.Rollback();
                    response.VentaExitosa = false;
                    response.Mensaje = "Error en el proceso de venta: " + ex.Message;
                }
            }

            return response;
        }

        private Electrodomestico BuscarElectrodomestico(SqlConnection connection, SqlTransaction transaction, int idElectrodomestico)
        {
            using (var cmd = connection.CreateCommand())
            {
                cmd.Transaction = transaction;
                cmd.CommandText = "SELECT * FROM ELECTRODOMESTICO WHERE idElectrodomestico = @id";
                cmd.Parameters.AddWithValue("@id", idElectrodomestico);

                using (var reader = cmd.ExecuteReader())
                {
                    if (reader.Read())
                    {
                        return new Electrodomestico
                        {
                            IdElectrodomestico = Convert.ToInt32(reader["idElectrodomestico"]),
                            Nombre = Convert.ToString(reader["nombre"]),
                            PrecioVenta = Convert.ToDecimal(reader["precioVenta"])
                        };
                    }
                }
            }

            return null;
        }

        private bool VerificarCreditoBanquito(string cedula, decimal monto, int plazoMeses)
        {
            var binding = new BasicHttpBinding();
            var endpoint = new EndpointAddress(BANQUITO_SOAP_URL);
            var factory = new ChannelFactory<BanquitoService>(binding, endpoint);
            BanquitoService client = factory.CreateChannel();

            try
            {
                var request = new EvaluacionCreditoRequest
                {
                    Cedula = cedula,
                    MontoElectrodomestico = monto,
                    PlazoMeses = plazoMeses
                };

                var response = client.EvaluarCredito(request);
                var channel = (IClientChannel)client;
                if (channel.State == CommunicationState.Faulted)
                {
                    channel.Abort();
                    factory.Abort();
                }
                else
                {
                    channel.Close();
                    factory.Close();
                }

                return response != null && response.CreditoAprobado == true;
            }
            catch
            {
                try
                {
                    ((IClientChannel)client).Abort();
                    factory.Abort();
                }
                catch
                {
                    // Ignorar errores de cierre
                }

                return false;
            }
        }
        private void CrearOActualizarCliente(SqlConnection connection, SqlTransaction transaction, string cedula, string nombre)
        {
            using (var checkCmd = connection.CreateCommand())
            {
                checkCmd.Transaction = transaction;
                checkCmd.CommandText = "SELECT COUNT(*) FROM CLIENTE WHERE cedula = @cedula";
                checkCmd.Parameters.AddWithValue("@cedula", cedula);

                var count = Convert.ToInt32(checkCmd.ExecuteScalar());
                if (count == 0)
                {
                    using (var insertCmd = connection.CreateCommand())
                    {
                        insertCmd.Transaction = transaction;
                        insertCmd.CommandText = "INSERT INTO CLIENTE (cedula, nombre) VALUES (@cedula, @nombre)";
                        insertCmd.Parameters.AddWithValue("@cedula", cedula);
                        insertCmd.Parameters.AddWithValue("@nombre", nombre);
                        insertCmd.ExecuteNonQuery();
                    }
                }
            }
        }

        private int? CrearFactura(SqlConnection connection, SqlTransaction transaction, VentaRequest request,
            decimal subtotal, decimal descuento, decimal total)
        {
            var estadoFactura = string.Equals(request.FormaPago, "EFECTIVO", StringComparison.OrdinalIgnoreCase)
                ? "PAGADA"
                : "APROBADA";

            using (var cmd = connection.CreateCommand())
            {
                cmd.Transaction = transaction;
                cmd.CommandText = "INSERT INTO FACTURA (cedula, fecha, formaPago, subtotal, descuento, total, estado) " +
                                  "OUTPUT INSERTED.idFactura " +
                                  "VALUES (@cedula, CAST(GETDATE() AS DATE), @formaPago, @subtotal, @descuento, @total, @estado)";

                cmd.Parameters.AddWithValue("@cedula", request.CedulaCliente);
                cmd.Parameters.AddWithValue("@formaPago", request.FormaPago);
                cmd.Parameters.AddWithValue("@subtotal", subtotal);
                cmd.Parameters.AddWithValue("@descuento", descuento);
                cmd.Parameters.AddWithValue("@total", total);
                cmd.Parameters.AddWithValue("@estado", estadoFactura);

                var result = cmd.ExecuteScalar();
                if (result != null && result != DBNull.Value)
                {
                    return Convert.ToInt32(result);
                }
            }

            return null;
        }

        private void CrearDetalleFactura(SqlConnection connection, SqlTransaction transaction, int idFactura,
            string idElectrodomestico, string cantidad, decimal precioUnitario, decimal subtotalLinea)
        {
            using (var cmd = connection.CreateCommand())
            {
                cmd.Transaction = transaction;
                cmd.CommandText =
                    "INSERT INTO DETALLE_FACTURA (idFactura, idElectrodomestico, cantidad, precioUnitario, subtotalLinea) " +
                    "VALUES (@idFactura, @idElectrodomestico, @cantidad, @precioUnitario, @subtotalLinea)";
                cmd.Parameters.AddWithValue("@idFactura", idFactura);
                cmd.Parameters.AddWithValue("@idElectrodomestico", int.Parse(idElectrodomestico));
                cmd.Parameters.AddWithValue("@cantidad", int.Parse(cantidad));
                cmd.Parameters.AddWithValue("@precioUnitario", precioUnitario);
                cmd.Parameters.AddWithValue("@subtotalLinea", subtotalLinea);
                cmd.ExecuteNonQuery();
            }
        }

        public List<Electrodomestico> ListarElectrodomesticos()
        {
            var electrodomesticos = new List<Electrodomestico>();

            using (var connection = DatabaseConnection.GetConnection())
            using (var cmd = connection.CreateCommand())
            {
                cmd.CommandText = "SELECT * FROM ELECTRODOMESTICO ORDER BY nombre";

                using (var reader = cmd.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        electrodomesticos.Add(new Electrodomestico
                        {
                            IdElectrodomestico = Convert.ToInt32(reader["idElectrodomestico"]),
                            Nombre = Convert.ToString(reader["nombre"]),
                            PrecioVenta = Convert.ToDecimal(reader["precioVenta"])
                        });
                    }
                }
            }

            return electrodomesticos;
        }
    }
}
